### PR TITLE
Add `raw.rescale`

### LINKED ISSUE #12111: RFC: should we add raw.rescale() ?
On [the forum](https://mne.discourse.group/t/troubles-in-show-bdf-files/7670/2) today, @cbrnr noted that for users whose data files aren't in Volts,

> Unfortunately, there is no easy-to-find rescale method, but you can use:
> `raw.apply_function(lambda x: x * 1e-6, picks="eeg")`

This seems a common enough need that I wonder if we should make it easier via a public method. Possible APIs:

1. `raw.rescale(scaler: float, picks=None)`
2. `raw.rescale(from: str = 'uV', to:str = 'V', picks='eeg')`

open to other API ideas (assuming folks think it's a good addition in the first place)

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #12111.

### CODE REVIEW

**Comment by cbrnr** [comment]
On a related note, could someone who uses Visual Studio Code try to debug my newly added test? Somehow, the debugger seems to be stuck when I (1) set a breakpoint in line 1070 and (2) then debug the test. Running the test does work though! Debugging tests work in other projects, so I was assuming that it's something specific to MNE. And yes, I already removed the venv and created a new one, basically using `uv venv; source .venv/bin/activate; uv pip install -e ".[test]"; uv pip install pyside6`. I also tried Python 3.13 and 3.12, and I also removed our custom pytest setup.


https://github.com/user-attachments/assets/2c5f18df-795a-4235-b65e-e854c8363bab

**Comment by drammock** [comment]
> could someone who uses Visual Studio Code try to debug my newly added test?

same result as you.  I seem to be hitting https://github.com/microsoft/debugpy/issues/1623

```
./mne/io/tests/test_raw.py::test_rescale Failed: [undefined]matplotlib._api.deprecation.MatplotlibDeprecationWarning: The interactive_bk attribute was deprecated in Matplotlib 3.9 and will be removed in 3.11. Use ``matplotlib.backends.backend_registry.list_builtin(matplotlib.backends.BackendFilter.INTERACTIVE)`` instead.
mne/io/tests/test_raw.py:1070: in test_rescale
    raw = read_raw_fif(raw_fname, preload=True)  # multiple channel types
<stringsource>:69: in cfunc.to_py.__Pyx_CFunc_b0409f__29_pydevd_sys_monitoring_cython_object__lParen__etc_to_py_4code_4line.wrap
    ???
_pydevd_sys_monitoring\\_pydevd_sys_monitoring_cython.pyx:1429: in _pydevd_sys_monitoring_cython._line_event
    ???
_pydevd_sys_monitoring\\_pydevd_sys_monitoring_cython.pyx:1471: in _pydevd_sys_monitoring_cython._internal_line_event
    ???
_pydevd_sys_monitoring\\_pydevd_sys_monitoring_cython.pyx:1278: in _pydevd_sys_monitoring_cython._stop_on_breakpoint
    ???
_pydevd_sys_monitoring\\_pydevd_sys_monitoring_cython.pyx:1906: in _pydevd_sys_monitoring_cython._do_wait_suspend
    ???
/home/drmccloy/.vscode/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydevd.py:2197: in do_wait_suspend
    keep_suspended = self._do_wait_suspend(thread, frame, event, arg, trace_suspend_type, from_this_thread, frames_tracker)
/home/drmccloy/.vscode/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydevd.py:2223: in _do_wait_suspend
    self._activate_gui_if_needed()
/home/drmccloy/.vscode/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydevd.py:1685: in _activate_gui_if_needed
    activate_function()
/home/drmccloy/.vscode/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydevd.py:1668: in <lambda>
    "matplotlib": lambda: activate_matplotlib(do_enable_gui),
/home/drmccloy/.vscode/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydev_ipython/matplotlibtools.py:99: in activate_matplotlib
    is_interactive = is_interactive_backend(backend)
/home/drmccloy/.vscode/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/pydev_ipython/matplotlibtools.py:59: in is_interactive_backend
    from matplotlib.rcsetup import interactive_bk, non_interactive_bk  # @UnresolvedImport
/opt/mambaforge/envs/mnedev/lib/python3.12/site-packages/matplotlib/_api/__init__.py:216: in __getattr__
    return props[name].__get__(instance)
/opt/mambaforge/envs/mnedev/lib/python3.12/site-packages/matplotlib/_api/deprecation.py:161: in __get__
    emit_warning()
/opt/mambaforge/envs/mnedev/lib/python3.12/site-packages/matplotlib/_api/deprecation.py:196: in emit_warning
    warn_deprecated(
/opt/mambaforge/envs/mnedev/lib/python3.12/site-packages/matplotlib/_api/deprecation.py:99: in warn_deprecated
    warn_external(warning, category=MatplotlibDeprecationWarning)
/opt/mambaforge/envs/mnedev/lib/python3.12/site-packages/matplotlib/_api/__init__.py:381: in warn_external
    warnings.warn(message, category, stacklevel)
E   matplotlib._api.deprecation.MatplotlibDeprecationWarning: The interactive_bk attribute was deprecated in Matplotlib 3.9 and will be removed in 3.11. Use ``matplotlib.backends.backend_registry.list_builtin(matplotlib.backends.BackendFilter.INTERACTIVE)`` instead.
```

**Comment by cbrnr** [comment]
Thanks @drammock. So what do we do? The issue is half a year old and it doesn't look like it's on the top of their to do list

**Comment by drammock** [comment]
> Thanks @drammock. So what do we do? The issue is half a year old and it doesn't look like it's on the top of their to do list

IDK.  I forked and patched debugpy locally, but haven't yet managed to get [the debugpy *extension*](https://github.com/microsoft/vscode-python-debugger) to use the local, patched copy.

In the meantime, if instead of a breakpoint you add a `1/0` or `raise RuntimeError` or similar, then the "right-click -> debug test" works as expected (for me at least)

**Comment by larsoner** [comment]
> Debugging tests work in other projects, so I was assuming that it's something specific to MNE. 

It probably is due to our treating all uncaught warnings as errors. If you add a line to our exceptions [here](https://github.com/mne-tools/mne-python/blob/730358c1e2e17baf0491c04e0c8269382a29c613/mne/conftest.py#L180) like:
```
    ignore:.*The interactive_bk attribute was deprecated:
```
Does it work? On at least get to another warning you could ignore?

**Comment by larsoner** [comment]
>    ignore:The interactive_bk attribute was deprecated.*:MatplotlibDeprecationWarning

I would naively expect this to be problematic because `MatplotlibDeprecationWarning` is not a fully qualified name or builtin, so the test suite shouldn't even run IIRC (might be what your error means, you might see it if you run `pytest` yourself?). You should need:
```
    ignore:The interactive_bk attribute was deprecated.*:matplotlib._api.deprecation.MatplotlibDeprecationWarning
```
Or similar.

**Comment by drammock** [comment]
> I would naively expect this to be problematic

ah, of course.

in the end I had to capture another similar warning too:

```py
    # debugpy uses deprecated matplotlib API
    ignore:The (non_)?interactive_bk attribute was deprecated.*:
```

**Comment by drammock** [comment]
@cbrnr I just pushed the fix to this branch. Debugging should hopefully work correctly for you now.

**Comment by mscheltienne** [comment]
I've been hitting this issue on VSCode for a while, ignoring the error in the conftest.py is my go-to solution. I think it's present in the conftest.py of MNE-LSL for instance.

**Comment by cbrnr** [comment]
Thanks everyone for your help, I am really happy that I can debug tests again! I've included the warning as suggested and expanded the examples prose a bit, so this should be ready.

**Review by drammock — CHANGES_REQUESTED** [review]

File: mne/io/base.py [hunk 1]
```diff
@@ -1500,6 +1500,67 @@ def resample(
...
+            if len(self.get_channel_types(unique=True)) == 1:
+                self.apply_function(lambda x: x * scale, channel_wise=False)
+            else:
+                raise ValueError(
+                    "If scale is a scalar, all channels must be of the same type. "
+                    "Consider passing a dict instead."
+                )
+        else:
+            for ch_type, ch_scale in scale.items():
+                if ch_type not in self.get_channel_types():
```
[the remark this answers is not in the document]
Comment (cbrnr) [reply mne/io/base.py:1555]: Good catch, I'll change this tomorrow!

[the remark this answers is not in the document]
Comment (cbrnr) [reply mne/io/base.py]: Done!

**Review by drammock — APPROVED** [review]
LGTM. My only hesitation is the param name `scale`, which is basically identical to what elsewhere we (I think?) consistently call `scalings`. To me `scale` is not obviously clearer (here or elsewhere) so its only advantage (I think?) is fewer characters. What do others think?

**Comment by drammock** [comment]
> I do not really have a preference, but `scalings` is plural, which is a bit awkward when rescaling with a single factor (e.g. `raw.rescale(scalings=1e-6)`).

yeah that occurred to me, but the converse is also true (singular param name when passing a dict with multiple scaling values)

**Comment by larsoner** [comment]
I think `scalings` is probably more consistent with what we do elsewhere

**Comment by cbrnr** [comment]
I'll change it to `scalings` then.

**Review by drammock — APPROVED** [review]

**Pull request merged**